File: Coding_Questions/JavaScript/ValidIPAddresses.js
// ---Valid IP Addresses---

const validIPAddresses = (string) => {
  const validIPArray = [];

  for (let i = 1; i < Math.min(string.length, 4); i++) {
    const ipAddressSectionsArray = ["", "", "", ""];

    ipAddressSectionsArray[0] = string.slice(0, i);

    if (!checkIPAddressSection(ipAddressSectionsArray[0])) {
      continue;
    }

    for (let c = i + 1; c < i + Math.min(string.length - i, 4); c++) {
      ipAddressSectionsArray[1] = string.slice(i, c);

      if (!checkIPAddressSection(ipAddressSectionsArray[1])) {
        continue;
      }

      for (let r = c + 1; r < c + Math.min(string.length - c, 4); r++) {
        ipAddressSectionsArray[2] = string.slice(c, r);
        ipAddressSectionsArray[3] = string.slice(r, string.length);

        if (
          checkIPAddressSection(ipAddressSectionsArray[2]) &&
          checkIPAddressSection(ipAddressSectionsArray[3])
        ) {
          validIPArray.push(ipAddressSectionsArray.join("."));
        }
      }
    }
  }

  return validIPArray;
};

const checkIPAddressSection = (section) => {
  return !(
    parseInt(section) > 255 ||
    (section[0] === "0" && section.length > 1)
  );
};

const checkIfListsAreEqual = (list1, list2) => {
  if (list1.length !== list2.length) {
    return false;
  }
  for (let i = 0; i < list1.length; i++) {
    if (list1[i] !== list2[i]) {
      return false;
    }
  }
  return true;
};
const input = "1921680";
const expected = [
  "1.9.216.80",
  "1.92.16.80",
  "1.92.168.0",
  "19.2.16.80",
  "19.2.168.0",
  "19.21.6.80",
  "19.21.68.0",
  "19.216.8.0",
  "192.1.6.80",
  "192.1.68.0",
  "192.16.8.0",
];
const actual = validIPAddresses(input);
console.log(checkIfListsAreEqual(actual, expected));

File: Coding_Questions/Python/ValidIPAddresses.py
# ---Valid IP Addresses---

from typing import List


def valid_ip_addresses(string: str) -> List[str]:
    valid_ips_list = list()

    for f_idx in range(1, 4):
        for s_idx in range(f_idx + 1, f_idx + 4):
            for t_idx in range(s_idx + 1, len(string)):
                potential_ip_address = string[:f_idx] + "." + \
                                       string[f_idx:s_idx] + "." + \
                                       string[s_idx:t_idx] + "." + \
                                       string[t_idx:]

                if check_valid_ip_address(potential_ip_address) is True:
                    valid_ips_list.append(potential_ip_address)

    return valid_ips_list


def check_valid_ip_address(string: str) -> bool:
    split_str = string.split(".")

    for section in split_str:
        full_number = ""
        for idx, number in enumerate(section):
            if idx == 0 and number == "0" and len(section) > 1:
                return False
            full_number += number

        if int(full_number) > 255:
            return False

    return True


def check_valid_ip_section(string: str) -> bool:
    return not (int(string) > 255 or (len(string) > 1 and string[0] == "0"))


def check_list_equality(list1: List[str], list2: List[str]) -> bool:
    return len(list1) == len(list2) and sorted(list1) == sorted(list2)


input1 = "1921680"
expected1 = [
    "1.9.216.80",
    "1.92.16.80",
    "1.92.168.0",
    "19.2.16.80",
    "19.2.168.0",
    "19.21.6.80",
    "19.21.68.0",
    "19.216.8.0",
    "192.1.6.80",
    "192.1.68.0",
    "192.16.8.0",
]
actual1 = valid_ip_addresses(input1)
print(f"EXPECTED:\t{expected1}")
print("------------------------")
print(f"ACTUAL:\t\t{actual1}")
assert check_list_equality(actual1, expected1)
